### ISSUE TITLE
Having more usage examples in documentation

### ISSUE THREAD

**Issue by TheMoonDawg** [gcanti/fp-ts#1320]
## 📖 Documentation

Hello! I'm relatively new to functional programming, so I've been trying to use this library with our TypeScript project. To be blunt, I'm still relatively bad at reading function signatures (but slowly improving)! A lot of the docs don't have usage examples which really trips me up at times. For example, I'm currently trying to insert a new value into a Map, and this is the API for that function:

```insertAt
Insert or replace a key/value pair in a map

Signature

export declare const insertAt: <K>(E: Eq<K>) => <A>(k: K, a: A) => (m: Map<K, A>) => Map<K, A>
Added in v2.0.0```

The function signature is pretty cryptic to me, so seeing a usage example goes a LONG way in helping me understand how it works. Is there a chance more usage examples could be added throughout the documentation to help newbies like me?

Thank you for this awesome library!

**Comment by SRachamim** [comment]
The documentation lacks descriptions and examples, you're right, but reading type signature are a bit fun :)

Let's take a look at the signature (I'll do a ELI5, so It may be overteaching):

```typescript
<K>(E: Eq<K>) => <A>(k: K, a: A) => (m: Map<K, A>) => Map<K, A>
```

1. First thing is to grasp the structure of the function: * -> * -> * -> *. i.e. It has three three invocations, so we can invoke it in one shout with `insertAt(*)(*)(*)`, or one at a time when it's, e.g.:

```typescript
  const f = insertAt(*);

const pipe = (
  someMap,
  f(*),
); 
```
Here, the first invocation creates an `f` function, the second invocation is on the second step on the pipe, while the third (and last) one will be done by the `pipe` function.

2. Second things is to dig the type parameters: We have a `K` and an `A`, and both can be anything! Easy.

3. First invocation is a `K` wrapped by an `Eq`. You will have to do know what is an [Eq](https://gcanti.github.io/fp-ts/modules/Eq.ts.html), or to research it. You can also see the [interface](https://gcanti.github.io/fp-ts/modules/Eq.ts.html#eq-interface).

Second invocation has to arguments: a `K` and an `A`.

Third (and last) invocation is a `Map<K, A>`. All functions on `fp-ts` are designed with the data-structure they operate on as the last invocation, so it'll be natural to `pipe` and `flow` them.

The return type is a `Map<K, A>`.

4. So we can derive that `K` is the type of the key in the map, and `A` is the type of the value in the map. Assuming our map is of type `Map<string, number>`, let's switch the `*` (`eqString` is an instance of `Eq` for `string`, i.e. `Eq<string>`):

```typescript
  const insertAtString = insertAt(eqString);

const pipe = (
  someMap,
  insertAtString('a', 1),
); 
```

**Comment by adworacz** [comment]
I was literally going to open an issue for this same thing when I found this, so thank you @TheMoonDawg, you're not alone.

As for @SRachamim's example - I'm glad that you enjoy reading type signatures. I admit, I'm getting better myself, but even with that improvement, finding example code is like a glass of cold water - shocking in how refreshing it is.

With examples like `fp-ts-std`, which documents (with examples) pretty much every method in the whole library, I think it's reasonable to aim for the same level of documentation for this library. Ex: https://samhh.github.io/fp-ts-std/modules/Array.ts.html

I'm certainly willing to put in work to this effect myself. My knowledge is still very limited with respect to this library, so the quality of my submissions will be questionable at times.

Perhaps it would be a good idea to update https://gcanti.github.io/fp-ts/guides/code-conventions such that all future submissions to the library require an associated example, as a way of putting the right foot forward?
